### PR TITLE
Add Code 32 Reader

### PR DESCRIPTION
**Pull request by stefanocali**
New extension of Reader Code 39. (Italian Pharmacode Code 32)

### CODE REVIEW

**Comment by ericblade** [comment]
Thanks for the contribution!  If you'd like, you are welcome to add yourself to the contributors list in package.json in this pull request.

Clearly this will need some glue to connect in with the rest of the lib, and I can have a look at what exactly is needed there.  We will also want some test cases, so if you can get some examples of the barcode as pictures in varying situations and angles, putting some jpgs of those, in /test/fixtures/code_32 would be really helpful.

I can generate barcode images, but it'd be better to have real life examples. :-)

Again, thanks!

**Comment by stefanocali** [comment]
Thanks to you! I uploaded everything you asked for.
I hope we can implement this code 32 soon! :)

File: src/decoder/barcode_decoder.js [hunk 1]
```diff
@@ -13,6 +13,7 @@ import UPCEReader from '../reader/upc_e_reader';
 import I2of5Reader from '../reader/i2of5_reader';
 import TwoOfFiveReader from '../reader/2of5_reader';
 import Code93Reader from '../reader/code_93_reader';
+import Code32Reader from '../reader/code_32_reader';
```
Comment (ericblade) [inline src/decoder/barcode_decoder.js:16]: I think if you add it to READERS here, at line 32, that might be all that you need to be able to reference it in Quagga.init()  and then we should be able to see some results?

Comment (ericblade) [reply src/decoder/barcode_decoder.js:16]: I'll try and pull this later today, i've been spending what little time i have to code on trying to get ean_reader to work in Typescript, but I think I can shift gears and give this a look

Comment (stefanocali) [reply src/decoder/barcode_decoder.js:16]: OK thanks! Let me know if everything goes well!

**Comment by ericblade** [comment]
OK, i'm apparently either still not a github expert, or it's still not possible to push a commit into other people's PRs.  :-)

I've added the tests in branch pr/stefanocali/164 .. if you pull that branch, then ```npm run build && npm run test``` then you can see that the tests fail.  The failure point appears to be the first line of Code32Reader.prototype._decode, whcih calls Code39Reader.prototype._decode.apply(this) ... which returns null.

My first thought, is of course it does, because it's code 39, not code 32 .. but then i remember that Code 32 is a *subset* of code 39, not a superset.  So, it should? work.  I think.  

My next thought is that the images might not be good enough resolution.  I just checked image-001.jpeg and that seems pretty good.  I'd be surprised if that's the problem.

I'm not sure how to proceed from there.  I kind of wish that we had the original authors around to tell us how he managed to build all of these readers. :-)

**Comment by stefanocali** [comment]
I'm trying to understand where the problem could be, I have not identified one on line number 9 ( code_32_reader.js ) there is to be changed {17} with {10} because the length of the code is 10 characters including the first one which is an A ...

**Comment by stefanocali** [comment]
Hi Eric, I made the changes and now with this code 32 image it works. It doesn't make me load the new version. Can you test it and insert it in the official version? Thanks, I await your news.

```
import Code39Reader from './code_39_reader';
import { BarcodePosition, Barcode } from './barcode_reader';

const patterns = {
    AEIO: /[AEIO]/g,
    AZ09: /[A-Z0-9]/,
};

const plaintextcode = [
    { "digit_code": 0, "character_code": "0" },
    { "digit_code": 1, "character_code": "1" },
    { "digit_code": 2, "character_code": "2" },
    { "digit_code": 3, "character_code": "3" },
    { "digit_code": 4, "character_code": "4" },
    { "digit_code": 5, "character_code": "5" },
    { "digit_code": 6, "character_code": "6" },
    { "digit_code": 7, "character_code": "7" },
    { "digit_code": 8, "character_code": "8" },
    { "digit_code": 9, "character_code": "9" },
    { "digit_code": 10, "character_code": "B" },
    { "digit_code": 11, "character_code": "C" },
    { "digit_code": 12, "character_code": "D" },
    { "digit_code": 13, "character_code": "F" },
    { "digit_code": 14, "character_code": "G" },
    { "digit_code": 15, "character_code": "H" },
    { "digit_code": 16, "character_code": "J" },
    { "digit_code": 17, "character_code": "K" },
    { "digit_code": 18, "character_code": "L" },
    { "digit_code": 19, "character_code": "M" },
    { "digit_code": 20, "character_code": "N" },
    { "digit_code": 21, "character_code": "P" },
    { "digit_code": 22, "character_code": "Q" },
    { "digit_code": 23, "character_code": "R" },
    { "digit_code": 24, "character_code": "S" },
    { "digit_code": 25, "character_code": "T" },
    { "digit_code": 26, "character_code": "U" },
    { "digit_code": 27, "character_code": "V" },
    { "digit_code": 28, "character_code": "W" },
    { "digit_code": 29, "character_code": "X" },
    { "digit_code": 30, "character_code": "Y" },
    { "digit_code": 31, "character_code": "Z" }
];

class Code32Reader extends Code39Reader {
    FORMAT = 'code_32';

    // TODO (this was todo in original repo, no text was there. sorry.)
    _checkChecksum(code: string) {
        return !!code;
    }

    _decode(row?: Array<number>, start?: BarcodePosition): Barcode | null {
        const result = super._decode(row, start);
        if (!result) {
            return null;
        }

        var code = result.code;

        if (!code) {
            return null;
        }

        code = code.replace(patterns.AEIO, '');

        if (!code.match(patterns.AZ09)) {
            if (ENV.development) {
                console.log('Failed AZ09 pattern code:', code);
            }
            return null;
        }

        if (!this._checkChecksum(code)) {
            return null;
        }

        var new_code = 0;
        var code_final;

        for (let i = 0; i <= code.length; i++) {
            let char_code = code.charAt(i);
            let search_digit_code = plaintextcode.find(e => e.character_code === char_code);

            if (typeof search_digit_code !== 'undefined') {
                let digit_code = search_digit_code.digit_code;

                let exponent_number = code.length - i - 1;

                digit_code = digit_code * Math.pow(32, exponent_number);
                new_code += digit_code;
            }
            
        }

        if (new_code.toString().length !== 9) {
            for (let n = 9; n > new_code.toString.length; n++) {
                code_final = "0"+new_code;
            }
        }

        code_final = "A"+new_code;

        result.code = code_final;
        return result;
    };
}

export default Code32Reader;
```

Link image code: https://barcodeguide.seagullscientific.com/Content/Symbologies/Code_32.htm

**Comment by ericblade** [comment]
hmm. i am having issues trying to pull this.. trying to work it out

File: src/reader/code_32.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                let exponent_number = code.length - i - 1;
+
+                digit_code = digit_code * Math.pow(32, exponent_number);
+                new_code += digit_code;
+            }
+            
+        }
+
+        if (new_code.toString().length !== 9) {
+            for (let n = 9; n > new_code.toString.length; n++) {
```
Comment (ericblade) [inline src/reader/code_32.ts]: looks like this should be new_code.toString().length .. also, it's an infinite loop, so something is definitely not right here.  I'm not sure what you're trying at, but this for loop is definitely not good

**Comment by stefanocali** [comment]
that loop cycle is used to add many 0's to the beginning of the string until the string is composed of 9 numbers.  

example 

if 2 numbers are missing to reach the length of 9 numbers it will add two 0. 

at the end of the cycle the character A is inserted at the beginning of the string

**Comment by ericblade** [comment]
ok, so, if the code you get is length 7, and the loop starts at 9, and you add one to n every time it loops, but the loop only terminates when n is no longer less than the string length, then the loop will go forever.  

I don't see any situation in which this isn't an infinite loop

**Comment by stefanocali** [comment]
I thought I had already modified it before the for loop. I updated the file.

**Comment by ericblade** [comment]
LGTM now, just going to pull it, run tests, in a little while.  Thank you so much!

**Pull request merged**